File: python/CyScat/Scattering_Code/scattering_coefficients_all.py
"""
scatteringcoefficientsall Code
Coder: Curtis Jin
Date: 2010/DEC/3rd Friday
Contact: jsirius@umich.edu
Description: Professor Michelson's version
           : Entries of the Scattering Matrix Generating Code
Translated to Python
"""

import numpy as np


def scatteringcoefficientsall(clocs, cmmaxs, period, lambda_wave, nmax, 
                              current_coefficients, up_down, d, sp):
    """
    Generate scattering coefficients for all cylinders
    
    Parameters:
    -----------
    clocs : ndarray
        Cylinder locations (Nx2 array)
    cmmaxs : ndarray
        Maximum mode numbers for each cylinder
    period : float
        Period of the structure
    lambda_wave : float
        Wavelength
    nmax : int
        Maximum mode number for plane waves
    current_coefficients : ndarray
        Current cylinder coefficients
    up_down : int
        Direction flag (1 for up, -1 for down)
    d : float
        Distance/thickness
    sp : dict
        SMatrix parameters
        
    Returns:
    --------
    scattering_coefficient_list : ndarray
        List of scattering coefficients (2*nmax+1 x 2)
    """
    scattering_coefficient_list = np.zeros((2*nmax + 1, 2), dtype=complex)
    max_mode_index = int(np.max(cmmaxs))
    
    for cm in range(-max_mode_index, max_mode_index + 1):
        # Find cylinders that have this mode
        cylinder_index = np.where(cmmaxs >= abs(cm))[0]
        
        # Calculate coefficient indices
        coefficient_index = []
        for j in range(len(cylinder_index)):
            cyl_idx = cylinder_index[j]
            idx = int(np.sum(2*cmmaxs[:cyl_idx] + 1) + cmmaxs[cyl_idx] + cm)
            coefficient_index.append(idx)
        
        coefficient_index = np.array(coefficient_index, dtype=int)
        
        # Add contribution from this mode
        scattering_coefficient_list += scatteringcoefficients_grid(
            clocs[cylinder_index, :], cm, period, lambda_wave, nmax,
            current_coefficients[coefficient_index], up_down, d, sp
        )
    
    return scattering_coefficient_list


def scatteringcoefficients_grid(clocs, cm, period, lambda_wave, nmax, 
                               current_coefficients, up_down, d, sp):
    """
    Calculate scattering coefficients on a grid
    
    Parameters:
    -----------
    clocs : ndarray
        Cylinder locations for selected cylinders
    cm : int
        Mode number
    period : float
        Period of the structure
    lambda_wave : float
        Wavelength
    nmax : int
        Maximum mode number
    current_coefficients : ndarray
        Current coefficients for selected cylinders
    up_down : int
        Direction flag
    d : float
        Distance/thickness
    sp : dict
        SMatrix parameters
        
    Returns:
    --------
    scattering_coefficient_list : ndarray
        Scattering coefficients (2*nmax+1 x 2)
    """
    index1 = sp['MiddleIndex'] - nmax
    index_end = sp['MiddleIndex'] + nmax
    index = np.arange(index1, index_end + 1)
    
    # Pre-calculate matrices
    kxs_index = sp['kxs'][index]
    kys_index = sp['kys'][index]
    angles_index = sp['Angles'][index]
    
    # Calculate dot products
    kxs_clocx = kxs_index @ clocs[:, 0:1].T
    kys_clocy = kys_index @ clocs[:, 1:2].T
    kys_clocy_minus_d = kys_index @ (clocs[:, 1:2] - d).T
    kys_clocy_plus_d = kys_index @ (clocs[:, 1:2] + d).T
    
    # Create angle mesh
    junk, angles_mesh = np.meshgrid(current_coefficients, angles_index)
    
    # Calculate coefficient
    # MATLAB: periodOVERkysCOEFF = sp.TwoOverPeriod ./ sp.kys(Index) * transpose(currentcoefficients);
    # This is: (scalar ./ column_vector) * row_vector → matrix of shape (len(kys), len(coeff))
    # Protect against division by zero for grazing incidence modes
    eps_ky = 1e-10
    kys_safe = kys_index.copy()
    near_zero_mask = np.abs(kys_safe) < eps_ky
    if np.any(near_zero_mask):
        kys_safe = np.where(near_zero_mask, eps_ky + 0j, kys_safe)
    period_over_kys_coeff = (sp['TwoOverPeriod'] / kys_safe.reshape(-1, 1)) @ current_coefficients.reshape(1, -1)
    
    scattering_coefficient_list = scatteringcoefficients_m_matrix(
        clocs, cm, period, lambda_wave, nmax, current_coefficients, up_down, d, sp,
        kxs_clocx, kys_clocy, kys_clocy_minus_d, kys_clocy_plus_d, 
        angles_mesh, period_over_kys_coeff
    )
    
    return scattering_coefficient_list


def scatteringcoefficients_m_matrix(clocs, cm, period, lambda_wave, nmax, 
                                   current_coefficient, up_down, d, sp,
                                   kxs_clocx, kys_clocy, kys_clocy_minus_d, 
                                   kys_clocy_plus_d, angles, period_over_kys_coeff):
    """
    Calculate scattering coefficients matrix
    
    Parameters:
    -----------
    All pre-calculated matrices and parameters
        
    Returns:
    --------
    s : ndarray
        Scattering coefficients [s11, s21] or [s12, s22]
    """
    # S11 & S21 Partition
    if up_down > 0:
        # S11 Partition
        exponent = kxs_clocx - kys_clocy + cm * (angles + np.pi)
        s11 = (-1)**cm * np.exp(1j * exponent) * period_over_kys_coeff
        s11 = np.sum(s11, axis=1, keepdims=True)
        
        # S21 Partition
        exponent = kxs_clocx + kys_clocy_minus_d - cm * (angles - np.pi)
        s21 = np.exp(1j * exponent) * period_over_kys_coeff
        s21 = np.sum(s21, axis=1, keepdims=True)
        
        s = np.hstack([s11, s21])
    
    # S22 & S12 Partition
    else:
        # S12 Partition
        exponent = kxs_clocx - kys_clocy_plus_d + cm * (angles + np.pi)
        s12 = (-1)**cm * np.exp(1j * exponent) * period_over_kys_coeff
        s12 = np.sum(s12, axis=1, keepdims=True)
        
        # S22 Partition
        exponent = kxs_clocx + kys_clocy - cm * (angles - np.pi)
        s22 = np.exp(1j * exponent) * period_over_kys_coeff
        s22 = np.sum(s22, axis=1, keepdims=True)
        
        s = np.hstack([s12, s22])
    
    return s

File: julia/CyScat/Scattering_Code/scattering_coefficients_all.jl
"""
scatteringcoefficientsall Code
Coder: Curtis Jin
Date: 2010/DEC/3rd Friday
Contact: jsirius@umich.edu
Description: Professor Michelson's version
           : Entries of the Scattering Matrix Generating Code

Julia translation - directly from MATLAB
"""

"""
    scatteringcoefficientsall(clocs, cmmaxs, period, lambda_wave, nmax,
                              current_coefficients, up_down, d, sp)

Generate scattering coefficients for all cylinders.
Exact translation from MATLAB scatteringcoefficientsall.m
"""
function scatteringcoefficientsall(clocs, cmmaxs, period, lambda_wave, nmax,
                                   current_coefficients, up_down, d, sp)
    # current_coefficients is already complex; use its element type directly
    CT = eltype(current_coefficients)
    scattering_coefficient_list = zeros(CT, 2*nmax + 1, 2)
    max_mode_index = Int(maximum(cmmaxs))

    for cm in -max_mode_index:max_mode_index
        # MATLAB: CylinderIndex = find(cmmaxs >= abs(cm));
        cylinder_index = findall(cmmaxs .>= abs(cm))

        # MATLAB: CoefficientIndex = zeros(length(CylinderIndex),1);
        coefficient_index = zeros(Int, length(cylinder_index))

        for j in 1:length(cylinder_index)
            # MATLAB: CoefficientIndex(j) = sum(2*cmmaxs(1:CylinderIndex(j)-1)+1)+cmmaxs(CylinderIndex(j))+1+cm;
            cyl_idx = cylinder_index[j]
            if cyl_idx == 1
                coefficient_index[j] = Int(cmmaxs[cyl_idx]) + 1 + cm
            else
                coefficient_index[j] = Int(sum(2 .* cmmaxs[1:cyl_idx-1] .+ 1)) + Int(cmmaxs[cyl_idx]) + 1 + cm
            end
        end

        # Add contribution from this mode
        scattering_coefficient_list .+= scatteringcoefficients_grid(
            clocs[cylinder_index, :], cm, period, lambda_wave, nmax,
            current_coefficients[coefficient_index], up_down, d, sp
        )
    end

    return scattering_coefficient_list
end


"""
    scatteringcoefficients_grid(clocs, cm, period, lambda_wave, nmax,
                                current_coefficients, up_down, d, sp)

Calculate scattering coefficients on a grid.
Exact translation from MATLAB scatteringcoefficientsGrid
"""
function scatteringcoefficients_grid(clocs, cm, period, lambda_wave, nmax,
                                     current_coefficients, up_down, d, sp)
    # MATLAB: Index1 = sp.MiddleIndex-nmax;
    # MATLAB: IndexEnd = sp.MiddleIndex+nmax;
    # MATLAB: Index = Index1:IndexEnd;
    index1 = sp["MiddleIndex"] - nmax
    index_end = sp["MiddleIndex"] + nmax
    index = index1:index_end

    # Ensure clocs is a matrix even for single cylinder
    if ndims(clocs) == 1
        clocs = reshape(clocs, 1, :)
    end

    # Ensure current_coefficients is a column vector
    current_coefficients = vec(current_coefficients)

    # Get the indexed arrays as column vectors (matching MATLAB)
    # In MATLAB, sp.kxs(Index) returns a column vector
    # We need to extract and reshape to column
    kxs_idx = vec(sp["kxs"][index, :])  # (n_modes,) column vector
    kys_idx = vec(sp["kys"][index, :])  # (n_modes,) column vector
    angles_idx = vec(sp["Angles"][index, :])  # (n_modes,) column vector

    # MATLAB: kxsclocx = sp.kxs(Index)*clocs(:,1)';
    # (n_modes,) * (1, n_cyl) -> need to reshape for proper matrix mult
    # In MATLAB: (7,1) * (1,3) = (7,3)
    # NOTE: Use transpose(), not ' (adjoint) to avoid complex conjugation
    kxs_clocx = kxs_idx * transpose(clocs[:, 1])

    # MATLAB: kysclocy = sp.kys(Index)*clocs(:,2)';
    kys_clocy = kys_idx * transpose(clocs[:, 2])

    # MATLAB: kysclocyminusd = sp.kys(Index)*(clocs(:,2)-d)';
    kys_clocy_minus_d = kys_idx * transpose(clocs[:, 2] .- d)

    # MATLAB: kysclocyplusd = sp.kys(Index)*(clocs(:,2)+d)';
    kys_clocy_plus_d = kys_idx * transpose(clocs[:, 2] .+ d)

    # MATLAB: [junk Angles] = meshgrid(currentcoefficients,sp.Angles(Index));
    # meshgrid(A, B) in MATLAB: X replicates A along rows, Y replicates B along columns
    # For Y (Angles): each column has sp.Angles(Index) values
    # Shape: (length(angles_idx), length(current_coefficients)) = (n_modes, n_cyl)
    n_modes = length(angles_idx)
    n_cyl = length(current_coefficients)
    angles_mesh = repeat(angles_idx, 1, n_cyl)  # Each column is angles_idx

    # MATLAB: periodOVERkysCOEFF = sp.TwoOverPeriod ./ sp.kys(Index) * transpose(currentcoefficients);
    # sp.TwoOverPeriod ./ sp.kys(Index) = scalar ./ (n_modes,) = (n_modes,)
    # (n_modes,) * (1, n_cyl) = (n_modes, n_cyl)
    # NOTE: Use transpose(), not ' (adjoint) to avoid complex conjugation of coefficients!
    period_over_kys = sp["TwoOverPeriod"] ./ kys_idx
    period_over_kys_coeff = period_over_kys * transpose(current_coefficients)

    scattering_coefficient_list = scatteringcoefficients_m_matrix(
        clocs, cm, period, lambda_wave, nmax, current_coefficients, up_down, d, sp,
        kxs_clocx, kys_clocy, kys_clocy_minus_d, kys_clocy_plus_d,
        angles_mesh, period_over_kys_coeff
    )

    return scattering_coefficient_list
end


"""
    scatteringcoefficients_m_matrix(...)

Calculate scattering coefficients matrix.
Exact translation from MATLAB scatteringcoefficientsmMatrix
"""
function scatteringcoefficients_m_matrix(clocs, cm, period, lambda_wave, nmax,
                                         current_coefficient, up_down, d, sp,
                                         kxs_clocx, kys_clocy, kys_clocy_minus_d,
                                         kys_clocy_plus_d, angles, period_over_kys_coeff)
    # S11 & S21 Partition
    if up_down > 0
        # MATLAB: exponent = kxsclocx - kysclocy + cm*(Angles+pi);
        exponent = kxs_clocx .- kys_clocy .+ cm .* (angles .+ π)

        # MATLAB: s11 = (-1).^cm .* exp(1i*exponent).*periodOVERkysCOEFF;
        s11 = ((-1.0)^cm) .* exp.(1im .* exponent) .* period_over_kys_coeff

        # MATLAB: s11 = sum(s11,2);
        s11 = sum(s11, dims=2)

        # MATLAB: exponent = kxsclocx + kysclocyminusd - cm*(Angles-pi);
        exponent = kxs_clocx .+ kys_clocy_minus_d .- cm .* (angles .- π)

        # MATLAB: s21 = exp(1i*exponent).*periodOVERkysCOEFF;
        s21 = exp.(1im .* exponent) .* period_over_kys_coeff

        # MATLAB: s21 = sum(s21,2);
        s21 = sum(s21, dims=2)

        # MATLAB: s = [s11,s21];
        s = hcat(vec(s11), vec(s21))

    # S22 & S12 Partition
    else
        # MATLAB: exponent = kxsclocx - kysclocyplusd + cm*(Angles+pi);
        exponent = kxs_clocx .- kys_clocy_plus_d .+ cm .* (angles .+ π)

        # MATLAB: s12 = (-1).^cm .* exp(1i*exponent).*periodOVERkysCOEFF;
        s12 = ((-1.0)^cm) .* exp.(1im .* exponent) .* period_over_kys_coeff

        # MATLAB: s12 = sum(s12,2);
        s12 = sum(s12, dims=2)

        # MATLAB: exponent = kxsclocx + kysclocy - cm*(Angles-pi);
        exponent = kxs_clocx .+ kys_clocy .- cm .* (angles .- π)

        # MATLAB: s22 = exp(1i*exponent).*periodOVERkysCOEFF;
        s22 = exp.(1im .* exponent) .* period_over_kys_coeff

        # MATLAB: s22 = sum(s22,2);
        s22 = sum(s22, dims=2)

        # MATLAB: s = [s12,s22];
        s = hcat(vec(s12), vec(s22))
    end

    return s
end
